src/parser/parse.ts

```diff
@@ -0,0 +1,204 @@
+import { UsefulToken } from '../lexer/tokens'
+import { TextField, IdentifierField, FieldValue, Field, Package } from './ipkg'
+import { tokensToString } from '../lexer/to-string'
+
+export type ParseSuccess<T> = { resultType: 'success'; value: T }
+export type ParseError = { resultType: 'error'; message: string }
+export type ParseResult<T> = ParseSuccess<T> | ParseError
+
+/**
+ * Parse the head of the package file.
+ *
+ * E.g.: `package PackageName`
+ */
+function parsePackageName(
+    tokens: Array<UsefulToken>,
+    pos: number,
+): ParseResult<{ name: string; newPos: number }> {
+    const packageIdent = tokens[pos]
+    const packageName = tokens[pos + 1]
+    if (
+        packageIdent &&
+        packageIdent.tokenType === 'ident' &&
+        packageIdent.identifier === 'package' &&
+        packageName &&
+        packageName.tokenType === 'ident'
+    ) {
+        return {
+            resultType: 'success',
+            value: { name: packageName.identifier, newPos: pos + 2 },
+        }
+    } else {
+        return {
+            resultType: 'error',
+            message: 'the ipkg file needs to start with a package description',
+        }
+    }
+}
+
+/**
+ * Return the proper field value from a token.
+ * Return `null` if it's not a valid token.
+ */
+function tokenValueToFieldValue(
+    token: UsefulToken,
+): TextField | IdentifierField | null {
+    switch (token.tokenType) {
+        case 'ident': {
+            return { fieldType: 'identifier', identifier: token.identifier }
+        }
+        case 'string-lit': {
+            return { fieldType: 'text', text: token.string }
+        }
+        default: {
+            return null
+        }
+    }
+}
+
+/**
+ * Parse a field value. The thing(s) after the "=".
+ */
+function parseFieldValue(
```

Comment:
There is some inconsistent usage of the term "field value": Sometimes (here for example) it means "all the things after the =" and thus potentially more than one item. Sometimes a single field can have multiple values. I would always use the plural and name this function `parseFieldValues` etc.